In [1]:
%matplotlib inline
import pandas as pd

Define display format

In [2]:
from IPython.core.display import HTML
css = open('style-table.css').read() + open('style-notebook.css').read()
HTML('<style>{}</style>'.format(css))

In [3]:
titles = pd.DataFrame.from_csv('data/titles.csv', index_col=None)
titles.head()

title  year
0       The Corruptor  1999
1          Nightshade  2004
2               Chut!  1972
3       Is That Nice?  1926
4  Tayna zolotoy gory  1985

In [4]:
cast = pd.DataFrame.from_csv('data/cast.csv', index_col=None)
cast.head()

title  year      name   type                character     n
0        Closet Monster  2015  Buffy #1  actor                  Buffy 4  31.0
1       Suuri illusioni  1985    Homo $  actor                   Guests  22.0
2   Battle of the Sexes  2017   $hutter  actor          Bobby Riggs Fan  10.0
3  Secret in Their Eyes  2015   $hutter  actor          2002 Dodger Fan   NaN
4            Steve Jobs  2015   $hutter  actor  1988 Opera House Patron   NaN

### How many movies are listed in the titles dataframe?

In [9]:
# all counts
titles.count()

title    244914
year     244914
dtype: int64

In [11]:
# unique counts
len(titles.title.value_counts())

223828

There are totally 244914 movies listed in the titles dataframe, among them there are totally 223828 unique titles.

### What are the earliest two films listed in the titles dataframe?

In [13]:
titles = titles.sort_values('year', ascending = False)

In [14]:
titles.head(2)

title  year
166442                 100 Years  2115
210275  The Zero Century: Maetel  2026

### How many movies have the title "Hamlet"?

In [17]:
titles[titles['title'] == 'Hamlet'].count()

title    20
year     20
dtype: int64

20 movies have the titles 'Hamlet'.

### How many movies are titled "North by Northwest"?

In [18]:
titles[titles['title'] == 'North by Northwest'].count()

title    1
year     1
dtype: int64

Only 1 movie is titled 'North by Northwest'.

### When was the first movie titled "Hamlet" made?

In [21]:
titles[titles['title'] == 'Hamlet'].sort_values('year')

title  year
105603  Hamlet  1910
174215  Hamlet  1911
118945  Hamlet  1913
26527   Hamlet  1921
10586   Hamlet  1948
127752  Hamlet  1954
58912   Hamlet  1964
108097  Hamlet  1969
23951   Hamlet  1973
140797  Hamlet  1976
34655   Hamlet  1987
193371  Hamlet  1990
126876  Hamlet  1996
69477   Hamlet  2000
42681   Hamlet  2009
170392  Hamlet  2011
162121  Hamlet  2014
125373  Hamlet  2015
64624   Hamlet  2016
101529  Hamlet  2017

The first movie titled "Hamlet" was made in 1910.

### List all of the "Treasure Island" movies from earliest to most recent.

In [22]:
titles[titles['title'] == 'Treasure Island'].sort_values('year')

title  year
73894   Treasure Island  1918
360     Treasure Island  1920
182290  Treasure Island  1934
190658  Treasure Island  1950
213127  Treasure Island  1972
25141   Treasure Island  1973
182997  Treasure Island  1985
209807  Treasure Island  1999

### How many movies were made in the year 1950?

In [23]:
titles[titles['year'] == 1950].count()

title    1113
year     1113
dtype: int64

1113 movies were made in the year 1950.

### How many movies were made in the year 1960?

In [24]:
titles[titles['year'] == 1960].count()

title    1549
year     1549
dtype: int64

1549 movies were made in the year 1950.

### How many movies were made from 1950 through 1959?

In [33]:
titles[(titles['year'] > 1949) & (titles['year'] < 1960)].count()

title    12934
year     12934
dtype: int64

12934 movies were made from 1950 through 1959.

### In what years has a movie titled "Batman" been released?

In [31]:
titles[titles['title'] == 'Batman']['year']

157682    1989
209326    1943
Name: year, dtype: int64

 A movie titled "Batman" has been released in the year 1943 and 1980.

### How many roles were there in the movie "Inception"?

In [37]:
cast[cast['title'] == 'Inception']['character'].count()

78

There are 78 characters in the movie "Inception".

### How many roles in the movie "Inception" are NOT ranked by an "n" value?

In [42]:
sum(cast[cast['title'] == 'Inception']['n'].isnull() == True)

27

27 roles in the movie "Inception" are not ranked by an 'n' value.

### But how many roles in the movie "Inception" did receive an "n" value?

In [43]:
sum(cast[cast['title'] == 'Inception']['n'].notnull() == True)

51

51 roles in the movie "Inception" are ranked by an 'n' value.

### Display the cast of "North by Northwest" in their correct "n"-value order, ignoring roles that did not earn a numeric "n" value.

In [48]:
cast[(cast['title'] == 'North by Northwest') & cast['n'].notnull()].sort_values('n', ascending = False)

title  year                  name     type  \
1384780  North by Northwest  1959             Ken Lynch    actor   
206471   North by Northwest  1959          Edward Binns    actor   
1508344  North by Northwest  1959         Patrick McVey    actor   
464736   North by Northwest  1959       Philip Coolidge    actor   
2295714  North by Northwest  1959          Les Tremayne    actor   
665768   North by Northwest  1959     Robert Ellenstein    actor   
1811387  North by Northwest  1959          Edward Platt    actor   
2447081  North by Northwest  1959         Adam Williams    actor   
1275337  North by Northwest  1959         Martin Landau    actor   
1695173  North by Northwest  1959           Philip Ober    actor   
3048272  North by Northwest  1959  Josephine Hutchinson  actress   
356707   North by Northwest  1959        Leo G. Carroll    actor   
3157232  North by Northwest  1959   Jessie Royce Landis  actress   
1457165  North by Northwest  1959           James Mason    actor   
3518333  North by Northwest  1959       Eva Marie Saint  actress   
871426   North by Northwest  1959            Cary Grant    actor   

                           character     n  
1384780  Charley - Chicago Policeman  16.0  
206471                Captain Junket  15.0  
1508344               Sergeant Flamm  14.0  
464736                     Dr. Cross  13.0  
2295714                   Auctioneer  12.0  
665768                         Licht  11.0  
1811387              Victor Larrabee  10.0  
2447081                     Valerian   9.0  
1275337                      Leonard   8.0  
1695173              Lester Townsend   7.0  
3048272                Mrs. Townsend   6.0  
356707                 The Professor   5.0  
3157232              Clara Thornhill   4.0  
1457165              Phillip Vandamm   3.0  
3518333                  Eve Kendall   2.0  
871426            Roger O. Thornhill   1.0

### Display the entire cast, in "n"-order, of the 1972 film "Sleuth".

In [50]:
cast[(cast['title'] == 'Sleuth') & (cast['year'] == 1972)].sort_values('n', ascending = False)

title  year                name     type  \
1448690  Sleuth  1972        Teddy Martin    actor   
2724553  Sleuth  1972  Eve (III) Channing  actress   
1465483  Sleuth  1972  John (II) Matthews    actor   
373876   Sleuth  1972      Alec Cawthorne    actor   
326773   Sleuth  1972       Michael Caine    actor   
1705394  Sleuth  1972    Laurence Olivier    actor   

                          character    n  
1448690      Police Constable Higgs  6.0  
2724553             Marguerite Wyke  5.0  
1465483  Detective Sergeant Tarrant  4.0  
373876            Inspector Doppler  3.0  
326773                  Milo Tindle  2.0  
1705394                 Andrew Wyke  1.0

### Now display the entire cast, in "n"-order, of the 2007 version of "Sleuth".

In [51]:
cast[(cast['title'] == 'Sleuth') & (cast['year'] == 2007)].sort_values('n', ascending = False)

title  year                 name     type          character    n
1805907  Sleuth  2007        Harold Pinter    actor        Man on T.V.  3.0
1293634  Sleuth  2007             Jude Law    actor               Milo  2.0
326774   Sleuth  2007        Michael Caine    actor             Andrew  1.0
260242   Sleuth  2007      Kenneth Branagh    actor  Other Man on T.V.  NaN
373877   Sleuth  2007  Alec (II) Cawthorne    actor  Inspector Doppler  NaN
2724552  Sleuth  2007    Eve (II) Channing  actress    Marguerite Wyke  NaN
3370424  Sleuth  2007    Carmel O'Sullivan  actress             Maggie  NaN

### How many roles were credited in the silent 1921 version of Hamlet?

In [52]:
cast[(cast['title'] == 'Hamlet') & (cast['year'] == 1921)]['character'].count()

9

9 roles were credited in the silent 1921 version of Hamlet.

### How many roles were credited in Branagh’s 1996 Hamlet?

In [53]:
cast[(cast['title'] == 'Hamlet') & (cast['year'] == 1996)]

title  year                    name     type  \
1731     Hamlet  1996              Riz Abbasi    actor   
95861    Hamlet  1996    Richard Attenborough    actor   
158642   Hamlet  1996     Simon Russell Beale    actor   
214473   Hamlet  1996         David (V) Blair    actor   
219554   Hamlet  1996           Brian Blessed    actor   
250773   Hamlet  1996      Christopher Bowles    actor   
260225   Hamlet  1996         Kenneth Branagh    actor   
272749   Hamlet  1996          Richard Briers    actor   
294652   Hamlet  1996          Michael Bryant    actor   
319501   Hamlet  1996            Peter Bygott    actor   
344365   Hamlet  1996       Kenneth W Caravan    actor   
500916   Hamlet  1996           Billy Crystal    actor   
518572   Hamlet  1996           Charles Daish    actor   
576147   Hamlet  1996        G?rard Depardieu    actor   
598618   Hamlet  1996          Reece Dinsdale    actor   
604339   Hamlet  1996                Ken Dodd    actor   
656817   Hamlet  1996             Rob Edwards    actor   
669232   Hamlet  1996             Jimmy Ellis    actor   
704270   Hamlet  1996        Nicholas Farrell    actor   
708468   Hamlet  1996              Ray Fearon    actor   
827066   Hamlet  1996            John Gielgud    actor   
990819   Hamlet  1996         Charlton Heston    actor   
1074691  Hamlet  1996           Ravil Isyanov    actor   
1082161  Hamlet  1996            Derek Jacobi    actor   
1210509  Hamlet  1996         Jeffery Kissoon    actor   
1316617  Hamlet  1996             Jack Lemmon    actor   
1400732  Hamlet  1996       Anthony Maddalena    actor   
1417360  Hamlet  1996         Michael Maloney    actor   
1489191  Hamlet  1996          Ian McElhinney    actor   
1549128  Hamlet  1996              John Mills    actor   
1556082  Hamlet  1996             Jimi Mistry    actor   
1588218  Hamlet  1996       Frank (II) Morgan    actor   
2040040  Hamlet  1996        Andrew Schofield    actor   
2052796  Hamlet  1996           Orlando Seale    actor   
2067657  Hamlet  1996            Rufus Sewell    actor   
2151526  Hamlet  1996           Timothy Spall    actor   
2154960  Hamlet  1996  John Spencer-Churchill    actor   
2221910  Hamlet  1996         Thomas Szekeres    actor   
2259675  Hamlet  1996                Ben Thom    actor   
2404091  Hamlet  1996          Don Warrington    actor   
2453099  Hamlet  1996          Robin Williams    actor   
2500066  Hamlet  1996               David Yip    actor   
2508377  Hamlet  1996             Jimmy Yuill    actor   
2737885  Hamlet  1996          Julie Christie  actress   
2814220  Hamlet  1996              Judi Dench  actress   
2835947  Hamlet  1996          Angela Douglas  actress   
2940033  Hamlet  1996           Yvonne Gidden  actress   
2999610  Hamlet  1996         Rosemary Harris  actress   
3119901  Hamlet  1996             Rowena King  actress   
3154270  Hamlet  1996               Sarah Lam  actress   
3451003  Hamlet  1996           Sian Radinger  actress   
3456114  Hamlet  1996          Melanie Ramsay  actress   
3728751  Hamlet  1996           Perdita Weeks  actress   
3751232  Hamlet  1996            Kate Winslet  actress   

                        character     n  
1731        Attendant to Claudius   1.0  
95861          English Ambassador   2.0  
158642         Second Gravedigger  37.0  
214473      Attendant to Claudius   3.0  
219554   Ghost of Hamlet's Father   4.0  
250773              Cadet in Play   NaN  
260225                     Hamlet   5.0  
272749                   Polonius   6.0  
294652                     Priest   7.0  
319501      Attendant to Claudius   8.0  
344365                Laertes Mob   NaN  
500916          First Gravedigger  10.0  
518572              Stage Manager  11.0  
576147                   Reynaldo  13.0  
598618               Guildenstern  14.0  
604339                     Yorick  15.0  
656817                   Lucianus  17.0  
669232             Old Fortinbras   NaN  
704270               

In [54]:
cast[(cast['title'] == 'Hamlet') & (cast['year'] == 1996)]['character'].count()

54

54 roles were credited in Branagh’s 1996 Hamlet.

### How many "Hamlet" roles have been listed in all film credits through history?

In [57]:
cast[cast['character'] == 'Hamlet']['character'].count()

96

96 "Hamlet" roles have been listed in all film credits through history.

### How many people have played an "Ophelia"?

In [61]:
cast[cast['character'] == 'Ophelia']['name'].value_counts()

Vanessa Leonard           2
Louise Beavers            2
Lucie Pohl                1
Marie Paquim              1
Marsha Warfield           1
Rosalinde Fuller          1
Anastasiya Vertinskaya    1
Maisie McCoy              1
Malaak Hattab             1
Adrienne Gould            1
Jennie Bushnell           1
Catherine (IV) Lord       1
Helen (IV) Wilson         1
Cailey Calisi             1
Ruby Dandridge            1
Bernadette Cuvalo         1
Pamela (II) Brown         1
Ophelia Shtruhl           1
Crystal Cordero           1
Rachel Grubb              1
Gertrude Elliot           1
Siobh?n F. Jess           1
Lucas Grabeel             1
Joanna Pickering          1
Cooper Harris             1
Josie Chisholm            1
Julia (III) Lee           1
Elise Kibler              1
Trisha Hitchcock          1
Mattie Peters             1
                         ..
Joanne Farrell            1
Simone Kirby              1
Kristi (V) Taylor         1
Sian Brooke               1
Justine Mooritz     

In [62]:
sum(cast[cast['character'] == 'Ophelia']['name'].value_counts())

117

117 people have played an "Ophelia". Two of them played two roles.

### How many people have played a role called "The Dude"?

In [63]:
sum(cast[cast['character'] == 'The Dude']['name'].value_counts())

19

19 people have played a role called "The Dude".

### How many people have played a role called "The Stranger"?

In [64]:
sum(cast[cast['character'] == 'The Stranger']['name'].value_counts())

212

212 people have played a role called "The Stranger".

### How many roles has Sidney Poitier played throughout his career?

In [67]:
cast[cast['name'] == 'Sidney Poitier']['character'].value_counts()

Virgil Tibbs                          3
FBI Deputy Director Carter Preston    1
Night Club Extra                      1
John Kane                             1
Steve Jackson                         1
Roy Parmenter                         1
Dr. Matt Younger                      1
Kimani Wa Karanja                     1
Manny Durrell                         1
Noah Cullen                           1
Marcus                                1
Homer Smith                           1
Walter Lee Younger                    1
Sgt. Eddie Towler                     1
Buck                                  1
Tommy Tyler                           1
Rau-Ru                                1
Clyde Williams                        1
Donald Crease                         1
Warren Stantin                        1
Shack Twala                           1
Robertson                             1
Alan Newell                           1
Porgy                                 1
Toller                                1


In [69]:
sum(cast[cast['name'] == 'Sidney Poitier']['character'].value_counts())

43

### How many roles has Judi Dench played?

In [70]:
sum(cast[cast['name'] == 'Judi Dench']['character'].value_counts())

55

### List the supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year.

In [72]:
cast[(cast['name'] == 'Cary Grant') & (cast['year'] >= 1940) & (cast['year'] < 1950) & (cast['n'] == 2)].sort_values('year')

title  year        name   type    character    n
871423  My Favorite Wife  1940  Cary Grant  actor   Nick Arden  2.0
871433    Penny Serenade  1941  Cary Grant  actor  Roger Adams  2.0

### List the leading roles that Cary Grant played in the 1940s in order by year.

In [79]:
cast[(cast['name'] == 'Cary Grant') & (cast['year'] >= 1940) & (cast['year'] < 1950) & (cast['n'] == 1)].sort_values('year')

title  year        name   type  \
871448               The Howards of Virginia  1940  Cary Grant  actor   
871407                       His Girl Friday  1940  Cary Grant  actor   
871450                The Philadelphia Story  1940  Cary Grant  actor   
871438                             Suspicion  1941  Cary Grant  actor   
871452                  The Talk of the Town  1942  Cary Grant  actor   
871429                 Once Upon a Honeymoon  1942  Cary Grant  actor   
871398                     Destination Tokyo  1943  Cary Grant  actor   
871422                             Mr. Lucky  1943  Cary Grant  actor   
871430                      Once Upon a Time  1944  Cary Grant  actor   
871390                  Arsenic and Old Lace  1944  Cary Grant  actor   
871425             None But the Lonely Heart  1944  Cary Grant  actor   
871424                         Night and Day  1946  Cary Grant  actor   
871427                             Notorious  1946  Cary Grant  actor   
871444      The Bachelor and the Bobby-Soxer  1947  Cary Grant  actor   
871445                     The Bishop's Wife  1947  Cary Grant  actor   
871421  Mr. Blandings Builds His Dream House  1948  Cary Grant  actor   
871402          Every Girl Should Be Married  1948  Cary Grant  actor   
871411                I Was a Male War Bride  1949  Cary Grant  actor   

                                   character    n  
871448                           Matt Howard  1.0  
871407                          Walter Burns  1.0  
871450                    C. K. Dexter Haven  1.0  
871438                      Johnnie Aysgarth  1.0  
871452                          Leopold Dilg  1.0  
871429                 Patrick 'Pat' O'Toole  1.0  
871398                         Capt. Cassidy  1.0  
871422  Joe Adams -posing as Joe Bascopolous  1.0  
871430                           Jerry Flynn  1.0  
871390                     Mortimer Brewster  1.0  
871425                            Ernie Mott  1.0  
871424                           Cole Porter  1.0  
871427                                Devlin  1.0  
871444                           Dick Nugent  1.0  
871445                                Dudley  1.0  
871421                         Jim Blandings  1.0  
871402                     Dr. Madison Brown  1.0  
871411                 Captain Henri Rochard  1.0

### How many roles were available for actors in the 1950s?

In [77]:
sum(cast[(cast['type'] == 'actor') & (cast['year'] >= 1950) & (cast['year'] < 1960)]['character'].value_counts())

157738

### How many roles were available for actresses in the 1950s?

In [78]:
sum(cast[(cast['type'] == 'actress') & (cast['year'] >= 1950) & (cast['year'] < 1960)]['character'].value_counts())

57733

### How many leading roles (n=1) were available from the beginning of film history through 1980?

In [82]:
sum(cast[(cast['n'] == 1) & (cast['year'] <= 1980)]['character'].value_counts())

65140

### How many non-leading roles were available through from the beginning of film history through 1980?

In [85]:
sum(cast[(cast['n'] > 1) & (cast['year'] <= 1980)]['character'].value_counts())

669320

In [86]:
sum(cast[(cast['n'] != 1) & (cast['year'] <= 1980)]['character'].value_counts())

1117667

### How many roles through 1980 were minor enough that they did not warrant a numeric "n" rank?

In [88]:
len(cast[(cast['n'].isnull()) & (cast['year'] <= 1980)]['character'])

448347